storage/parquet/datasource.go

```diff
@@ -417,39 +435,151 @@ func (rs *RecordStream) RunWorkerInternal(ctx context.Context, tx storage.StateT
 
 	batch := make([]*execution.Record, 0)
 	for i := 0; i < rs.batchSize; i++ {
-		record, err := rs.readRecordFromFile()
+		recordValues, err := rs.readRecordFromFile()
 		if err == execution.ErrEndOfStream {
 			break
 		} else if err != nil {
 			return errors.Wrap(err, "couldn't read record from parquet file")
 		}
 
-		aliasedRecord := make(map[octosql.VariableName]octosql.Value)
-		for k, v := range record {
-			if str, ok := v.(string); ok {
-				parsed, err := time.Parse(time.RFC3339, str)
-				if err == nil {
-					v = parsed
-				}
-			}
+		octoValues := make([]octosql.Value, len(rs.columns))
+		for i, v := range recordValues {
 			if int96, ok := v.(parquet.Int96); ok {
 				v = int(binary.LittleEndian.Uint64(int96[:8]))
 			}
-			aliasedRecord[octosql.NewVariableName(fmt.Sprintf("%s.%s", rs.alias, k))] = octosql.NormalizeType(v)
-		}
+			// Info here: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md
+			if rs.columnsLogicalTypes[i] != nil && v != nil {
+				var convert func(interface{}) interface{}
+
+				logicalType := rs.columnsLogicalTypes[i]
+				switch {
+				case logicalType.STRING != nil:
+					convert = func(v interface{}) interface{} {
+						return string(v.([]byte))
+					}
+				case logicalType.DECIMAL != nil:
+					convert = func(v interface{}) interface{} {
+						switch v := v.(type) {
+						case int32:
+							return float64(v) * math.Pow10(int(-logicalType.DECIMAL.Scale))
+						case int64:
+							return float64(v) * math.Pow10(int(-logicalType.DECIMAL.Scale))
+						case []byte:
+							var value float64
+							// This reads the unsigned value.
+							for i := range v {
+								j := len(v) - i - 1
+								value += float64(v[j]) * math.Pow(2, float64(8*i))
+							}
+							// This applies two's complement to get a signed one.
+							if v[0]&0b10000000 == 0b10000000 {
+								value -= math.Pow(2, float64(8*len(v)))
+							}
+							return value * math.Pow10(int(-logicalType.DECIMAL.Scale))
+						default:
+							return v
+						}
+					}
+				case logicalType.DATE != nil:
+					convert = func(v interface{}) interface{} {
+						epochStart := time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)
+						return epochStart.Add(time.Hour * 24 * time.Duration(v.(int32)))
+					}
+				case logicalType.TIME != nil:
+					convert = func(v interface{}) interface{} {
+						if logicalType.TIME.Unit.IsSetMILLIS() {
+							return time.Millisecond * time.Duration(v.(int32))
+						} else if logicalType.TIME.Unit.IsSetMICROS() {
+							return time.Microsecond * time.Duration(v.(int64))
+						}
+						return v
+					}
+				case logicalType.TIMESTAMP != nil:
+					convert = func(v interface{}) interface{} {
+						value := v.(int64)
+						if logicalType.TIMESTAMP.IsAdjustedToUTC {
+							var seconds int64
+							var nanoseconds int64
+							if logicalType.TIMESTAMP.Unit.IsSetMILLIS() {
+								seconds = value / 1000
```

Comment:
Isn't it better to use constants like `Time.second` and so on instead of raw values like 1000, 10^6 ?

Reply:
done, though I'm not that happy with the end result